src/registry/artifactory.rs

```diff
@@ -26,191 +27,166 @@ pub struct Artifactory {
+    pub async fn download(&self, dependency: Dependency) -> miette::Result<Package> {
+        let artifact_url = resolve_url(dependency.clone())?;
 
-        ensure!(status.is_success(), "Failed to ping artifactory");
+        tracing::debug!("Hitting download URL: {artifact_url}");
 
-        tracing::debug!("Pinging artifactory succeeded");
+        let response = self.new_request(Method::GET, artifact_url).send().await?;
 
-        Ok(())
+        convert_to_package(response)
+            .await
+            .wrap_err_with(|| format!("failed to download dependency {}", dependency.package))
     }
 
-    /// Downloads a package from artifactory
-    pub async fn download(&self, dependency: Dependency) -> eyre::Result<Package> {
-        ensure!(
-            dependency.manifest.version.comparators.len() == 1,
-            "{} uses unsupported semver comparators",
-            dependency.package
-        );
-
-        let version = dependency
-            .manifest
-            .version
-            .comparators
-            .first()
-            .wrap_err("internal error")?;
-
-        ensure!(
-            version.op == semver::Op::Exact && version.minor.is_some() && version.patch.is_some(),
-            "artifactory only support pinned dependencies"
-        );
+    /// Publishes a package to artifactory
+    pub async fn publish(&self, package: Package, repository: String) -> miette::Result<()> {
+        let artifact_uri: Url = format!(
+            "{}/{}/{}/{}-{}.tgz",
+            self.registry,
+            repository,
+            package.name(),
+            package.name(),
+            package.version(),
+        )
+        .parse()
+        .into_diagnostic()
+        .wrap_err(miette!(
+            "unexpected error: failed to construct artifact URL"
+        ))?;
+
+        let _ = self
+            .new_request(Method::PUT, artifact_uri)
+            .body(package.tgz.clone())
+            .send()
+            .await?;
 
-        let version = format!(
-            "{}.{}.{}{}",
-            version.major,
-            version.minor.wrap_err("internal error")?,
-            version.patch.wrap_err("internal error")?,
-            if version.pre.is_empty() {
-                "".to_owned()
-            } else {
-                format!("-{}", version.pre)
-            }
+        tracing::info!(
+            ":: published {}/{}@{}",
+            repository,
+            package.name(),
+            package.version()
         );
 
-        let artifact_url: Url = {
-            let path = dependency.manifest.registry.path().to_owned();
+        Ok(())
+    }
+}
 
-            let mut url = dependency.manifest.registry.clone();
-            url.set_path(&format!(
-                "{}/{}/{}/{}-{}.tgz",
-                path,
-                dependency.manifest.repository,
-                dependency.package,
-                dependency.package,
-                version
-            ));
+struct RequestBuilder(reqwest::RequestBuilder);
```

Comment:
why do we have this, this doesnt add any logic right?

Reply:
We discussed this, this is the Request wrapper that makes it so we can't make requests without validating the response.